src/NuGet.Server.Core/Infrastructure/ServerPackageRepository.cs

```diff
@@ -509,18 +538,56 @@ private void FileSystemChanged(object sender, FileSystemEventArgs e)
                 // TODO: invalidating *all* packages for every nupkg change under this folder seems more expensive than it should.
                 // Recommend using e.FullPath to figure out which nupkgs need to be (re)computed.
 
-                ClearCache();
+                await ClearCacheAsync(CancellationToken.None);
             }
         }
-        
-        private IDisposable LockAndSuppressFileSystemWatcher()
+
+        private async Task<DisposableSemphoreSlim> LockAsync(CancellationToken token)
+        {
+            var handle = new DisposableSemphoreSlim(_syncLock);
+            await handle.WaitAsync(token);
+            return handle;
+        }
+
+        private async Task<SupressedFileSystemWatcher> LockAndSuppressFileSystemWatcherAsync(CancellationToken token)
         {
-            return new SupressedFileSystemWatcher(this);
+            var handle = new SupressedFileSystemWatcher(this);
+            await handle.WaitAsync(token);
+            return handle;
+        }
+
+        private class DisposableSemphoreSlim : IDisposable
```

Comment:
Sounds fine. The intend is that `Dispose` releases the semaphore instead of what `Semaphore.Dispose` does.